### PR TITLE
GH-40819: [Java] Adding Spotless to Algorithm module

### PR DESCRIPTION
**Pull request by vibhatha**
### Rationale for this change

Adding code style and formatting options for Algorithm module. 

### What changes are included in this PR?

Code formatting spotless plugin has been added. 

### Are these changes tested?

Yes, but doesn't involve test cases, the plugin itself corrects. 

### Are there any user-facing changes?

No
* GitHub Issue: apache/arrow#40819
* GitHub Issue: #40819

### CODE REVIEW

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java [hunk 1]
```diff
@@ -1,108 +1,108 @@
...
-        // an equal element is found,
-        // continue to go right-ward
-        ret = mid;
-        low = mid + 1;
-      }
-    }
-    return ret;
-  }
-}
+/*
```
Comment (vibhatha) [inline java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java:1]: This change is a bit odd, I cannot see a formatting here, though locally the spotless gives a warning about the file. 

cc @danepitkin

**Comment by vibhatha** [comment]
@github-actions crossbow submit preview-docs

File: java/algorithm/pom.xml [hunk 2]
```diff
@@ -50,4 +55,68 @@
...
+  <!-- configure license for java files -->
+  <includes>
+    <include>**/*.java</include>
+  </includes>
+  <licenseHeader>
+    <file>${maven.multiModuleProjectDirectory}/java/spotless/asf-java.license</file>
+    <delimiter>package</delimiter>
+  </licenseHeader>
+</format>
+              </formats>
```
Comment (danepitkin) [inline java/algorithm/pom.xml:93]: nit: is formatting incorrect here?

File: java/algorithm/pom.xml [hunk 3]
```diff
@@ -50,4 +55,68 @@
 
   <build>
   </build>
+
+  <profiles>
+    <profile>
+      <id>spotless-jdk11+</id>
```
Comment (danepitkin) [inline java/algorithm/pom.xml:61]: Do we need a `spotless-jdk8`?

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java (hunk 1 above)
Comment (vibhatha) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java:1]: Good point, I did compare it with the main branch, it shows no change. That's the odd part. I need to check this again.

Comment (lidavidm) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java:1]: \r\n -> \n

Comment (lidavidm) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java:1]: @kou is exploring pre-commit (https://github.com/apache/arrow/issues/40417) which has checks for this sort of thing. Would Spotless be able to integrate with pre-commit?

Comment (vibhatha) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java:1]: > \r\n -> \n
This is acceptable?

Comment (vibhatha) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java:1]: @lidavidm I think we should be able to integrate it. We could create a bash script which does `mvn spotless:apply` and format the code. Shall we investigate in a separate PR?

File: java/algorithm/pom.xml (hunk 3 above)
Comment (vibhatha) [reply java/algorithm/pom.xml:61]: I assume we still do have some JDK 8 developers? That's why I wanted to leave it here, but if it is not a very strong case, we can definitely remove it. WDYT?

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java (hunk 1 above)
Comment (vibhatha) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java:1]: Issue created: https://github.com/apache/arrow/issues/40854

File: java/algorithm/pom.xml (hunk 2 above)
Comment (kou) [reply java/algorithm/pom.xml:93]: Should we add a formatter for XML too? :-)

File: java/algorithm/pom.xml (hunk 3 above)
Comment (lidavidm) [reply java/algorithm/pom.xml:61]: I don't want to have two separate versions of the formatter. Only build/test needs to work on JDK8.

File: java/algorithm/pom.xml (hunk 2 above)
Comment (lidavidm) [reply java/algorithm/pom.xml:93]: Yes, if there's a pre-commit plugin for XML that would be nice.

File: docs/source/developers/java/development.rst [hunk 4]
```diff
@@ -110,9 +110,66 @@ integration tests, you would do:
...
+The current Java code styles are configured as follows:
+
+- Indent: Tabs & spaces (2 spaces per tab)
+- Google Java Format: Reformats Java source code to comply with `Google Java Style`_.
+- Configure license headers for Java & XML files
+
+Java code style is checked by `Spotless`_ during the build, and the continuous integration build will verify
+that changes adhere to the style guide.
+
+.. code-block:: xml
```
Comment (lidavidm) [inline docs/source/developers/java/development.rst:122]: I don't see what this code block is for?

File: docs/source/developers/java/development.rst [hunk 5]
```diff
@@ -110,9 +110,66 @@ integration tests, you would do:
...
+
+Automatically fixing code style issues
+--------------------------------------
+
+- You can also just check the style without building the project with `mvn spotless:check`.
+- The Java code style can be corrected from the command line by using the following commands: `mvn spotless:apply`.
+
+.. code-block:: bash
+
+    user@machine repo % mvn spotless:check
```
Comment (lidavidm) [inline docs/source/developers/java/development.rst:150]: Was this copied from their docs? Can we just provide our own example or omit this?

Comment (vibhatha) [reply docs/source/developers/java/development.rst:150]: I adopted some of the things from the older PR. Let me verify. Still an early draft. I will clean up before marking this ready for review. Sorry about this.

File: java/algorithm/pom.xml (hunk 2 above)
Comment (kou) [reply java/algorithm/pom.xml:93]: We may be able to use https://github.com/prettier/plugin-xml via Prettier.

@vibhatha Could you open a new issue for this?

File: java/algorithm/pom.xml [hunk 6]
```diff
@@ -50,4 +54,74 @@
...
+      <id>spotless</id>
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>com.diffplug.spotless</groupId>
+            <artifactId>spotless-maven-plugin</artifactId>
+            <version>${spotless.version}</version>
```
Comment (laurentgo) [inline java/algorithm/pom.xml:69]: Shoudn't this be managed at the top level pom.xml in `pluginManagement` if we are planning to enable it for all modules?

File: java/algorithm/pom.xml [hunk 7]
```diff
@@ -20,6 +20,10 @@
   <name>Arrow Algorithms</name>
   <description>(Experimental/Contrib) A collection of algorithms for working with ValueVectors.</description>
 
+  <properties>
+    <spotless.version>2.42.0</spotless.version>
```
Comment (laurentgo) [inline java/algorithm/pom.xml:24]: This version does not work with Java 8

File: java/algorithm/pom.xml [hunk 8]
```diff
@@ -50,4 +54,74 @@
...
+                  </includes>
+                  <licenseHeader>
+                    <file>${maven.multiModuleProjectDirectory}/java/spotless/asf-java.license</file>
+                    <delimiter>package</delimiter>
+                  </licenseHeader>
+                </format>
+              </formats>
+              <java>
+                <googleJavaFormat>
+                  <version>1.9</version>
```
Comment (laurentgo) [inline java/algorithm/pom.xml:95]: `1.9` is quite old but also is not compatible with Java 8 (I think the last Java 8 version is `1.7`)

File: java/algorithm/pom.xml [hunk 9]
```diff
@@ -50,4 +54,74 @@
...
+                  <style>GOOGLE</style>
+                </googleJavaFormat>
+              </java>
+              <pom>
+                <indent>
+                  <tabs>true</tabs>
```
Comment (laurentgo) [inline java/algorithm/pom.xml:101]: Not sure why having a rule to use tab indentation + a rule for space indentation + sortPom (which also handles indentation with default being 2 space per indentation level). Isn't `sortPom` enough?

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java (hunk 1 above)
Comment (laurentgo) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java:1]: The issue with this file are the line terminators (this is a Windows text file basically)

File: java/algorithm/pom.xml (hunk 7 above)
Comment (lidavidm) [reply java/algorithm/pom.xml:24]: IMO, it's OK for this to only be Java 11+. The only thing that needs to work right now on Java 8 is build and test, not necessarily static analysis.

Comment (laurentgo) [reply java/algorithm/pom.xml:24]: For people still using Java 8 to build Arrow, it means they may not notice a formatting issue (or apply a fix for it) unless they submit a pull request or use a different java version. Workable but not ideal

Comment (vibhatha) [reply java/algorithm/pom.xml:24]: Earlier I added a separate version for Java 8, but we settled to keep just one. @lidavidm should we add it back?

**Comment by vibhatha** [comment]
@laurentgo thanks for the feedback. I added a few quick comments for now, and I will address them accordingly.

File: java/algorithm/pom.xml (hunk 7 above)
Comment (laurentgo) [reply java/algorithm/pom.xml:24]: Could we use spotless [2.30.0](https://www.atlassian.com/agile/project-management/epics-stories-themes) until the build uses a newer java version?

Comment (vibhatha) [reply java/algorithm/pom.xml:24]: I can definitely try that.

File: java/algorithm/pom.xml (hunk 8 above)
Comment (lidavidm) [reply java/algorithm/pom.xml:95]: Well, which is it? Either both should be compatible or neither should be compatible. If you are going to use an older version of Spotless then also use an older version here.

**Comment by laurentgo** [comment]
@vibhatha I'd like to contribute to the effort. Is there something I can do to help you?

File: java/algorithm/pom.xml (hunk 2 above)
Comment (laurentgo) [reply java/algorithm/pom.xml]: This is addressed by #41173 I believe

**Comment by vibhatha** [comment]
@laurentgo that would be wonderful. I haven't push this PR in a while. We do have similar issues per module. Also the idea is to add code formatting and xml formatting.

File: java/algorithm/pom.xml (hunk 2 above)
Comment (vibhatha) [reply java/algorithm/pom.xml]: cc @kou

File: java/algorithm/pom.xml (hunk 6 above)
Comment (vibhatha) [reply java/algorithm/pom.xml:69]: We are planning to go module by module and probably after adding all we should move it to the top level?

**Comment by vibhatha** [comment]
I am going to close this PR and re-open a new one.

File: java/algorithm/pom.xml (hunk 2 above)
Comment (kou) [reply java/algorithm/pom.xml]: Can we integrate spotless with our `.pre-commit-config.yaml`?
https://github.com/apache/arrow/blob/main/.pre-commit-config.yaml

We can detect format error in our lint CI https://github.com/apache/arrow/blob/main/.github/workflows/dev.yml#L38 by integrating it with pre-commit.

Comment (vibhatha) [reply java/algorithm/pom.xml]: Sure @kou I will take a look.

**Pull request closed without merging**